samza-core/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseTableReadFunction.java

```diff
@@ -32,15 +37,41 @@
     implements TableReadFunction<String, V> {
   private static final Logger LOGGER = LoggerFactory.getLogger(CouchbaseTableReadFunction.class);
 
-  private void initial() {
-    cluster = CouchbaseCluster.create(clusterNodes);
-    cluster.authenticate(username, password);
-    bucket = cluster.openBucket(bucketName);
+  public CouchbaseTableReadFunction(Class<V> valueClass) {
+    super(valueClass);
+    LOGGER.info(String.format("Read function for bucket %s initialized successfully", bucketName));
   }
 
   @Override
-  public CompletableFuture<V> getAsync(String s) {
-    return CompletableFuture.supplyAsync(() -> (V) bucket.get(s));
+  public CompletableFuture<V> getAsync(String key) {
+    CompletableFuture<V> getFuture = new CompletableFuture<>();
+    SingleSubscriber<Document> subscriber = new SingleSubscriber<Document>() {
+      @Override
+      public void onSuccess(Document v) {
+        if (v == null) {
+          getFuture.complete(null);
+        }
+        if (JsonDocument.class.isAssignableFrom(valueClass)) {
+          getFuture.complete((V) v);
+        } else {
+          getFuture.complete(valueSerde.fromBytes((byte[]) v.content()));
+        }
+      }
+
+      @Override
+      public void onError(Throwable error) {
+        throw new SamzaException(String.format("Failed to get key %s", key), error);
```

Comment:
Shouldn't we mark future failed?
Are you sure we can throw here?

Reply:
I should wrap the SamzaException with future.completeExceptionally. Will change it.